### PR TITLE
Adding the windows logged_in_users virtual table

### PR DESCRIPTION
**Pull request by PoppySeedPlehzr**
This adds initial support for the logged in user, with a few items remaining that still need to be figured out. I haven't been able to find documentation or info on getting the PID affiliated with a console session on Windows, and I haven't yet experimented much with the IPv4 RDP sessions for parsing out remote hosts.

### CODE REVIEW

**Comment by osqueryer** [comment]
:-1: The commit 33b80a8cef9bfb2523929a12a8cde0ef1f22d094 (Job results: [228](https://jenkins.osquery.io/job/osqueryPullRequestOSX/228/)) failed one or more tests (**macOS/OS X**).

**Comment by PoppySeedPlehzr** [comment]
ok to test

File: osquery/tables/system/windows/logged_in_users.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ */
+
+#define _WIN32_DCOM
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <Wtsapi32.h>
+#include <winsock2.h>
+
+#include <osquery/core.h>
+#include <osquery/core/conversions.h>
```
Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: Use the `"osquery/core/conversions.h"` local search method in a block of includes below the public APIs.

File: osquery/tables/system/windows/logged_in_users.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+#include <osquery/filesystem/fileops.h>
```
Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: Same.

File: osquery/tables/system/windows/logged_in_users.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      WTS_CURRENT_SERVER_HANDLE, &level, 0, &pSessionInfo, &count);
+
+  if (res == 0) {
+    return results;
+  }
+
+  for (size_t i = 0; i < count; i++) {
+    Row r;
+
+    char* sessionInfo;
```
Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: Initialize to `nullptr`.

File: osquery/tables/system/windows/logged_in_users.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    unsigned long bytesRet = 0;
+    res = WTSQuerySessionInformation(WTS_CURRENT_SERVER_HANDLE,
+                                     pSessionInfo[i].SessionId,
+                                     WTSSessionInfo,
+                                     &sessionInfo,
+                                     &bytesRet);
+    if (res == 0) {
+      VLOG(1) << "Error querying WTS Session Information, Error Code ("
```
Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: nit, do you need to CAP Session Information Error Code?

Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: nit, do you need "error code" at all?

File: osquery/tables/system/windows/logged_in_users.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+              << GetLastError() << ")";
+      continue;
+    }
+    auto wtsSession = (PWTSINFO)sessionInfo;
```
Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: Is it possible this is still `nullptr`?

File: osquery/tables/system/windows/logged_in_users.cpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    FILETIME utcTime = {0};
+    unsigned long long unixTime = 0;
+    utcTime.dwLowDateTime = wtsSession->ConnectTime.LowPart;
+    utcTime.dwHighDateTime = wtsSession->ConnectTime.HighPart;
+    if (utcTime.dwLowDateTime != 0 || utcTime.dwHighDateTime != 0) {
+      unixTime = filetimeToUnixtime(utcTime);
+    }
+    r["time"] = INTEGER(unixTime);
+
+    char* clientInfo;
```
Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: Same.

File: osquery/tables/system/windows/logged_in_users.cpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    bytesRet = 0;
+    res = WTSQuerySessionInformation(WTS_CURRENT_SERVER_HANDLE,
+                                     pSessionInfo[i].SessionId,
+                                     WTSClientInfo,
+                                     &clientInfo,
+                                     &bytesRet);
+    if (res == 0) {
+      VLOG(1) << "Error querying WTS Session Information, Error Code ("
```
Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: Same.

File: osquery/tables/system/windows/logged_in_users.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      r["host"] = std::to_string(wtsClient->ClientAddress[0]) + "." +
+                  std::to_string(wtsClient->ClientAddress[1]) + "." +
+                  std::to_string(wtsClient->ClientAddress[2]) + "." +
+                  std::to_string(wtsClient->ClientAddress[3]);
+    } else if (wtsClient->ClientAddressFamily == AF_INET6) {
+      // TODO: IPv6 addresses are given as an array of byte values.
+      r["host"] = SQL_TEXT(wtsClient->ClientAddress);
+    }
+
+    r["pid"] = -1;
```
Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: Unfortunately we still need the boilerplate `INTEGER(-1)` :/

File: osquery/tables/system/windows/logged_in_users.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    results.push_back(r);
+  }
+
+  if (pSessionInfo != nullptr) {
+    WTSFreeMemoryEx(WTSTypeSessionInfoLevel1, pSessionInfo, count);
```
Comment (theopolis) [inline osquery/tables/system/windows/logged_in_users.cpp]: What about `clientInfo`?

**Comment by PoppySeedPlehzr** [comment]
I've implemented the requested changes, but now I'm seeing a small memory leak somewhere in the code. I'm trying to get some tooling to work with a Debug release to try and find this before proceeding. I've got `WTSFreeMemory` calls for each of the `WTSQuerySessionInformation` calls, so I'm fairly certain everything is being freed correctly..

**Comment by theopolis** [comment]
If you post your changes I can take a look too (but the best I can offer is reading the associated MSDNs) :)

**Comment by PoppySeedPlehzr** [comment]
@theopolis will do :)

I've been spending time trying to automate rolling out the Debug versions of all of the libs, This is like the 20th time I've wanted to have full Debug versions for VS Debugging :|

**Comment by PoppySeedPlehzr** [comment]
Ok, I got a debugging release up and I am less convinced there's a memory leak, I think I was noticing jumps in memory due to sqlite taking in new data, and assumed it was something wrong with the APIs? Either way, I think this is looking good now.

**Pull request merged**